Remove `IORuntime` from `allRuntimes` on shutdown

RafalSierkiewicz opened the pull request:
The PR originated from question on discord https://discord.com/channels/632277896739946517/632278585700384799/1254741830042652772 

Whenever the IORuntime is created through `apply` method it is being added to the `allRuntimes` collection.
Unfortunately it seems that is not removed anywhere. 

This PR tries to remove the created `IORuntime` from `allRuntimes` collection. 
I tried to keep the api as close as it was, but as the `allRuntimes` collection require `runtime` and the `hashCode` to be supplied to remove function I struggled a bit. 

I am open for comments and suggestions. Let me know which branch I should target, not sure if change is backward and forward compatible

Commit:
Remove IORuntime from allRuntimes collection

Commit:
Remove scala util chaining as it is not available on scala 2.12

Commit:
Run formatting after second commit

djspiewak reviewed (changes requested):
I think it would make more sense to do this in the same local scope where we *add* to `allRuntimes`, both because it avoids this parameter juggle in the main class, but also because it keeps the logic a bit more localized. The forward references should be resolvable with some type ascription and strategic use of the `lazy` modifier.

Commit:
Remove IORuntime in shutdown hook

RafalSierkiewicz commented:
> I think it would make more sense to do this in the same local scope where we _add_ to `allRuntimes`, both because it avoids this parameter juggle in the main class, but also because it keeps the logic a bit more localized. The forward references should be resolvable with some type ascription and strategic use of the `lazy` modifier.

Yes, I totally I agree. I was trying to achieve it, but somehow realized that might be hard 🙈 I changed it to use `lazy` as well as `def` and added test

djspiewak commented on core/shared/src/main/scala/cats/effect/unsafe/IORuntime.scala:
```diff
@@ -70,12 +70,13 @@ object IORuntime extends IORuntimeCompanionPlatform {
     val fiberMonitor = FiberMonitor(compute)
     val unregister = registerFiberMonitorMBean(fiberMonitor)
-    val unregisterAndShutdown = () => {
+    def unregisterAndShutdown: () => Unit = () => {
```
Moderate pedantry:

```suggestion
    def unregisterAndShutdown(): Unit = {
```

RafalSierkiewicz replied:
I had to revert the change because compilation is failing in such a solution 
`An unapplied 0-arity method was eta-expanded (due to the expected type () => Unit), rather than applied to `()`.
[error] Write unregisterAndShutdown() to invoke method unregisterAndShutdown, or change the expected type.`

RafalSierkiewicz committed:
Apply change suggestion

Co-authored-by: Daniel Spiewak <djspiewak@gmail.com>

Commit:
Revert "Apply change suggestion"

This reverts commit 12b6c54d7c79a707b8328315762e93c2aacb016b.
